pyvisa_py/protocols/hislip.py

```diff
@@ -250,463 +567,179 @@ def initialize(self, version=(1, 0), vendor_id=b"xx", sub_address=b"hislip0"):
-                    self.payload_remaining = self.get_data_header()
-
-            request_size = min(self.payload_remaining, max_len - bytes_recvd)
-            receive_exact_into(self.sock, view[:request_size])
-            self.payload_remaining -= request_size
-            bytes_recvd += request_size
-            view = view[request_size:]
-
-        if bytes_recvd > max_len:
-            raise MemoryError("scribbled past end of recv_buffer")
-
-        return recv_buffer
+        send_msg(self._sync, "DeviceClearComplete", feature_bitmap, 0)
+        response = DeviceClearAcknowledge(self._sync)
+        return response.feature_bitmap
+
+    def trigger(self) -> None:
+        """send a Trigger packet on the sync channel"""
+        send_msg(self._sync, "Trigger", self._rmt, self._message_id)
+        self._rmt = 0
+        self._last_message_id = self._message_id
+        self._message_id = (self._message_id + 2) & 0xFFFF_FFFF
+
+    def _send_data_packet(self, payload: bytes) -> None:
+        """send a Data packet on the sync channel"""
+        send_msg(self._sync, "Data", self._rmt, self._message_id, payload)
+        self._rmt = 0
+        self._last_message_id = self._message_id
+        self._message_id = (self._message_id + 2) & 0xFFFF_FFFF
+
+    def _send_data_end_packet(self, payload: bytes) -> None:
+        """send a DataEnd packet on the sync channel"""
+        send_msg(self._sync, "DataEnd", self._rmt, self._message_id, payload)
+        self._rmt = 0
+        self._last_message_id = self._message_id
+        self._message_id = (self._message_id + 2) & 0xFFFF_FFFF
+
+    def fatal_error(self, error: str, error_message: str = "") -> None:
+        err_msg = (error_message or error).encode()
+        send_msg(self._sync, "FatalError", FATALERRORCODE[error], 0, err_msg)
+
+    def error(self, error: str, error_message: str = "") -> None:
+        err_msg = (error_message or error).encode()
+        send_msg(self._sync, "Error", ERRORCODE[error], 0, err_msg)
+
+
+# def rxdecode(header):
+#     (
+#         prologue,
+#         msg_type,
+#         control_code,
+#         message_parameter,
+#         payload_length,
+#     ) = struct.unpack(HEADER_FORMAT, header)
+#
+#     msg_type = MESSAGETYPE_STR[msg_type]
+#     print(
+#         f"Rx: {prologue=}, "
+#         f"{msg_type=}, "
+#         f"{control_code=}, "
+#         f"{message_parameter=}, "
+#         f"{payload_length=}"
+#     )
+
+
+# def txdecode(header, payload=b""):
+#     (
+#         prologue,
+#         msg_type,
+#         control_code,
+#         message_parameter,
+#         payload_length,
+#     ) = struct.unpack(HEADER_FORMAT, header)
+#
+#     msg_type = MESSAGETYPE_STR[msg_type]
+#     print(
+#         f"Tx: {prologue=}, "
+#         f"{msg_type=}, "
+#         f"{control_code=}, "
+#         f"{message_parameter=}, "
+#         f"{payload_length=}, "
+#         f"{len(payload)=}, "
+#         f"{bytes(payload[:20]).decode('iso-8859-1')!r}"
+#     )
```

Comment:
Are those worth keeping around ? If so please add a comment explaining their purpose.

Reply:
I've added a comment above those routines, but left them commented out in order to support Python versions <= 3.7.

They were really useful in figuring out why I couldn't talk to the Keysight scope after some minor changes.  I had left out the network order format specifier in a call to struct.pack().  So when I set the max message size to 16MB, it looked like I was setting it to 4GB.  Except the Keysight scope has a bug where it throws away the upper 32 bits of the 64-bit unsigned value.  So it decided I had set the max message size to 0, and it kept sending me packets with zero length payloads.

Once I realized that it was ignoring the upper 32 bits, I couldn't resist seeing if it was interpreting the remaining bits as signed or unsigned.  Sure enough, the scope application crashes if you give it a value with bit 31 set.